src/cache/inmemory/fragmentRegistry.ts

```
@@ -0,0 +1,167 @@
+  }
+
+  private cacheUnaryMethod<TName extends keyof Pick<FragmentRegistry,
+    | "lookup"
+    | "transform"
+    | "findFragmentSpreads"
+  >>(name: TName) {
+    const registry = this;
+    const originalMethod = FragmentRegistry.prototype[name];
+    return wrap(function () {
+      return originalMethod.apply(registry, arguments);
+    }, {
+      makeCacheKey: arg => arg,
+    });
+  }
+
+  public lookup(fragmentName: string): FragmentDefinitionNode | null {
+    return this.registry[fragmentName] || null;
+  }
+
+  public transform<D extends DocumentNode>(document: D): D {
+    const defined = new Map<string, FragmentDefinitionNode>();
+    getFragmentDefinitions(document).forEach(def => {
+      defined.set(def.name.value, def);
+    });
+
+    const unbound = new Set<string>();
+    const enqueue = (spreadName: string) => {
+      if (!defined.has(spreadName)) {
+        unbound.add(spreadName);
+      }
+    };
+
+    const enqueueChildSpreads = (node: ASTNode) => Object.keys(
+      this.findFragmentSpreads(node)
+    ).forEach(enqueue);
+
+    enqueueChildSpreads(document);
+
+    const missing: string[] = [];
+    const map: FragmentMap = Object.create(null);
+
+    // This Set forEach loop can be extended during iteration by adding
+    // additional strings to the unbound set.
+    unbound.forEach(fragmentName => {
+      const knownFragmentDef = defined.get(fragmentName);
+      if (knownFragmentDef) {
+        enqueueChildSpreads(map[fragmentName] = knownFragmentDef);
+      } else {
+        missing.push(fragmentName);
+        const def = this.lookup(fragmentName);
+        if (def) {
+          enqueueChildSpreads(map[fragmentName] = def);
+        }
+      }
+    });
+
+    if (missing.length) {
+      const defsToAppend: FragmentDefinitionNode[] = [];
+      missing.forEach(name => {
+        const def = map[name];
+        if (def) {
+          defsToAppend.push(def);
+        } else {
+          // TODO Warn? Error?
```

If we end up here does that mean we have fragment spreads that don't have a matching fragment definition defined globally or locally? If so, I'm leaning towards an error (with a nice explanatory error message mentioning the fragment name); might as well fail fast if that's the case.